### PR TITLE
Expose Roslyn OOP as a limited LSP server supporting code search via extended LSP

### PR DESCRIPTION
**Pull request by olegtk**
This prototype exposes RoslynCodeAnalysis service hub service as a limited LSP server supporting only extended (streaming) version of LSP's workspace symbol search. This allows Roslyn to support VS search on par with GoTo.

More context: VS Search (former Quick Launch) public extensibility is LSP only, it supports both standard workspace/symbol and extended workspace/beginSymbol, workspace/publishSymbol, workspace/completeSymbol (not yet documented). Standard LSP's workspace/symbol doesn't support streaming of results yet (expected in next version via https://github.com/microsoft/vs-streamjsonrpc/issues/139), that's why until it does we need to extend LSP with workspace/beginSymbol.

This implementation is reusing AbstractNavigateToSearchServic, just like GoTo, but unlike GoTo it uses PrimaryWorkspace as LSP doesn't support workspace synchronization.

Initialization of this Roslyn LSP server is left to VS LSP client for now, it will initialize it when user starts interacting with VS Search (based on "Capability" metadata).

### CODE REVIEW

**Comment by olegtk** [comment]
cc @gundermanc

**Comment by CyrusNajmabadi** [comment]
Let me know when this is ready for review.  Thanks! :)

**Comment by heejaechang** [comment]
@olegtk thank you oleg. so if we want to test this new feature, which VS drop should we use? or should we wait until some vs part checked in to v.next?

**Comment by gundermanc** [comment]
@heejaechang there's a PR that needs to be checked into the LSP client for VS. I can meet with you in person to test this until it's in.

**Comment by olegtk** [comment]
> @olegtk thank you oleg. so if we want to test this new feature, which VS drop should we use? or should we wait until some vs part checked in to v.next?

As @gundermanc pointed out this PR depends on changes in Microsoft.VisualStudio.LanguageServer.Protocol.Extensions.dll first. Once they are in and it's published (Roslyn already depends on it) you should be able to use master builds for testing this PR.

File: src/VisualStudio/CSharp/Impl/LanguageServerClient/CSharpLanguageServerClient.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+{
+    [ContentType(ContentTypeNames.CSharpContentType)]
+    [Export(typeof(ILanguageClient))]
+    [ExportMetadata("Capability", "WorkspaceStreamingSymbolProvider")]
+    internal class CSharpLanguageServerClient : ILanguageClient
+    {
+        /// <summary>
+        /// Gets the name of the language client (displayed to the user).
+        /// </summary>
+        public string Name => "C# Language Server Client";
```
Comment (gundermanc) [inline src/VisualStudio/CSharp/Impl/LanguageServerClient/CSharpLanguageServerClient.cs]: IIRC this is shown to the user on exception. Do we need to localize?

**Comment by gundermanc** [comment]
TODO: NullReferenceException that needs to be fixed prior to check in:

```
>	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Remote.RemoteHostClientExtensions.TryGetRemoteHostClientAsync(Microsoft.CodeAnalysis.Workspace workspace, System.Threading.CancellationToken cancellationToken) Line 104	C#
 	Microsoft.VisualStudio.LanguageServices.CSharp.dll!Microsoft.VisualStudio.LanguageServices.CSharp.LanguageServerClient.CSharpLanguageServerClient.GetHostGroupIdAsync(System.Threading.CancellationToken cancellationToken) Line 66	C#
 	Microsoft.VisualStudio.LanguageServices.CSharp.dll!Microsoft.VisualStudio.LanguageServices.CSharp.LanguageServerClient.CSharpLanguageServerClient.ActivateAsync(System.Threading.CancellationToken token) Line 56	C#
 	Microsoft.VisualStudio.LanguageServer.Client.Implementation.dll!Microsoft.VisualStudio.LanguageServer.Client.RemoteLanguageClientInstance.InitializeAsync(string workspaceLocation, System.Threading.CancellationToken token)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start<System.__Canon>(ref System.__Canon stateMachine)	Unknown
 	Microsoft.VisualStudio.LanguageServer.Client.Implementation.dll!Microsoft.VisualStudio.LanguageServer.Client.RemoteLanguageClientInstance.InitializeAsync(string workspaceLocation, System.Threading.CancellationToken token)	Unknown
 	Microsoft.VisualStudio.LanguageServer.Client.Implementation.dll!Microsoft.VisualStudio.LanguageServer.Client.RemoteLanguageClientInstance.ActivateLanguageClientAsync(string workspaceLocation, System.Threading.CancellationToken token)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<bool>.Start<System.__Canon>(ref System.__Canon stateMachine)	Unknown
 	Microsoft.VisualStudio.LanguageServer.Client.Implementation.dll!Microsoft.VisualStudio.LanguageServer.Client.RemoteLanguageClientInstance.ActivateLanguageClientAsync(string workspaceLocation, System.Threading.CancellationToken token)	Unknown
 	Microsoft.VisualStudio.LanguageServer.Client.Implementation.dll!Microsoft.VisualStudio.LanguageServer.Client.RemoteLanguageClientInstance.InitializeStartLanguageServerTask.AnonymousMethod__0()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.InvokeMoveNext(object stateMachine)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.OutputAsyncCausalityEvents.AnonymousMethod__0()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.ContinuationWrapper.Invoke()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.TaskAwaiter.OutputWaitEtwEvents.AnonymousMethod__0()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.ContinuationWrapper.Invoke()	Unknown
 	mscorlib.dll!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action action, bool allowInlining, ref System.Threading.Tasks.Task currentTask)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.FinishContinuations()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.FinishStageThree()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task<System.__Canon>.TrySetResult(System.__Canon result)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<string>.SetResult(string result)	Unknown
 	Microsoft.VisualStudio.LanguageServer.Client.Implementation.dll!Microsoft.VisualStudio.LanguageServer.Client.RemoteLanguageClientInstance.GetLocationAsync()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.InvokeMoveNext(object stateMachine)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.OutputAsyncCausalityEvents.AnonymousMethod__0()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.ContinuationWrapper.Invoke()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.TaskAwaiter.OutputWaitEtwEvents.AnonymousMethod__0()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.ContinuationWrapper.Invoke()	Unknown
 	mscorlib.dll!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action action, bool allowInlining, ref System.Threading.Tasks.Task currentTask)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.FinishContinuations()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.FinishStageThree()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task<System.__Canon>.TrySetResult(System.__Canon result)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<Microsoft.VisualStudio.Shell.Interop.IVsSolution>.SetResult(Microsoft.VisualStudio.Shell.Interop.IVsSolution result)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.AsyncLazy<Microsoft.VisualStudio.Shell.Interop.IVsSolution>.GetValueAsync.AnonymousMethod__0()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.InvokeMoveNext(object stateMachine)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.OutputAsyncCausalityEvents.AnonymousMethod__0()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.ContinuationWrapper.Invoke()	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.AwaitExtensions.ExecuteContinuationSynchronouslyAwaiter<System.__Canon>.OnCompleted.AnonymousMethod__5_0(System.Threading.Tasks.Task<System.__Canon> _, object s)	Unknown
 	mscorlib.dll!System.Threading.Tasks.ContinuationTaskFromResultTask<Microsoft.VisualStudio.Shell.Interop.IVsSolution>.InnerInvoke()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.Execute()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.ExecutionContextCallback(object obj)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.ExecuteWithThreadLocal(ref System.Threading.Tasks.Task currentTaskSlot)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.ExecuteEntry(bool bPreventDoubleExecution)	Unknown
 	mscorlib.dll!System.Threading.Tasks.ThreadPoolTaskScheduler.TryExecuteTaskInline(System.Threading.Tasks.Task task, bool taskWasPreviouslyQueued)	Unknown
 	mscorlib.dll!System.Threading.Tasks.TaskScheduler.TryRunInline(System.Threading.Tasks.Task task, bool taskWasPreviouslyQueued)	Unknown
 	mscorlib.dll!System.Threading.Tasks.TaskContinuation.InlineIfPossibleOrElseQueue(System.Threading.Tasks.Task task, bool needsProtection)	Unknown
 	mscorlib.dll!System.Threading.Tasks.StandardTaskContinuation.Run(System.Threading.Tasks.Task completedTask, bool bCanInlineContinuationTask)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.FinishContinuations()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.FinishStageThree()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task<System.__Canon>.TrySetResult(System.__Canon result)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<Microsoft.VisualStudio.Shell.Interop.IVsSolution>.SetResult(Microsoft.VisualStudio.Shell.Interop.IVsSolution result)	Unknown
 	Microsoft.VisualStudio.LanguageServer.Client.Implementation.dll!Microsoft.VisualStudio.LanguageServer.Client.RemoteLanguageServiceBroker.GetSolutionServiceAsync()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.InvokeMoveNext(object stateMachine)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.OutputAsyncCausalityEvents.AnonymousMethod__0()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.ContinuationWrapper.Invoke()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.TaskAwaiter.OutputWaitEtwEvents.AnonymousMethod__0()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.ContinuationWrapper.Invoke()	Unknown
 	mscorlib.dll!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action action, bool allowInlining, ref System.Threading.Tasks.Task currentTask)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.FinishContinuations()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.FinishStageThree()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task<System.__Canon>.TrySetResult(System.__Canon result)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<object>.SetResult(object result)	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.AsyncServiceProvider.QueryServiceAsync.AnonymousMethod__0()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.InvokeMoveNext(object stateMachine)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.OutputAsyncCausalityEvents.AnonymousMethod__0()	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncMethodBuilderCore.ContinuationWrapper.Invoke()	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.VsTaskLibraryHelper.CreateTaskBody.AnonymousMethod__0(Microsoft.VisualStudio.Shell.Interop.IVsTask task, Microsoft.VisualStudio.Shell.Interop.IVsTask[] parents)	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.VsManagedTaskBody.DoWork(Microsoft.VisualStudio.Shell.Interop.IVsTask pTask, uint dwCount, Microsoft.VisualStudio.Shell.Interop.IVsTask[] pParentTasks, out object pResult)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.GetCallbackForMultipleParent.AnonymousMethod__0(System.Threading.Tasks.Task<object>[] _) Line 1137	C#
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.GetCallbackForSingleParent.AnonymousMethod__0(System.Threading.Tasks.Task<object> internalTask) Line 1178	C#
 	mscorlib.dll!System.Threading.Tasks.ContinuationResultTaskFromResultTask<object, object>.InnerInvoke()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.Execute()	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.ExecutionContextCallback(object obj)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.ExecuteWithThreadLocal(ref System.Threading.Tasks.Task currentTaskSlot)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.ExecuteEntry(bool bPreventDoubleExecution)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()	Unknown
 	mscorlib.dll!System.Threading.ThreadPoolWorkQueue.Dispatch()	Unknown
 	mscorlib.dll!System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()	Unknown
```

File: src/VisualStudio/CSharp/Impl/LanguageServerClient/CSharpLanguageServerClient.cs (hunk 1 above)
Comment (olegtk) [reply src/VisualStudio/CSharp/Impl/LanguageServerClient/CSharpLanguageServerClient.cs]: yes, this one has to be localized

**Comment by gundermanc** [comment]
I don't have permissions to push directly to the repo so I've fixed up @olegtk's PR to work with the final LSP protocol bits here: https://github.com/dotnet/roslyn/pull/37359.

**Pull request closed without merging**